components/drop-zone/index.js

```diff
@@ -0,0 +1,229 @@
+		window.removeEventListener( 'drop', this.onDrop );
+		window.removeEventListener( 'dragenter', this.toggleDraggingOverDocument );
+		window.removeEventListener( 'dragleave', this.toggleDraggingOverDocument );
+		window.removeEventListener( 'mouseup', this.resetDragState );
+
+		this.disconnectMutationObserver();
+	}
+
+	resetDragState() {
+		const { isDraggingOverDocument, isDraggingOverElement } = this.state;
+		if ( ! ( isDraggingOverDocument || isDraggingOverElement ) ) {
+			return;
+		}
+
+		this.setState( {
+			isDraggingOverDocument: false,
+			isDraggingOverElement: false,
+		} );
+	}
+
+	toggleMutationObserver() {
+		this.disconnectMutationObserver();
+
+		if ( this.state.isDraggingOverDocument ) {
+			this.observer = new window.MutationObserver( this.detectNodeRemoval );
+			this.observer.observe( document.body, {
+				childList: true,
+				subtree: true,
+			} );
+		}
+	}
+
+	disconnectMutationObserver() {
+		if ( ! this.observer ) {
+			return;
+		}
+
+		this.observer.disconnect();
+		delete this.observer;
+	}
+
+	detectNodeRemoval( mutations ) {
+		mutations.forEach( ( mutation ) => {
+			if ( ! mutation.removedNodes.length ) {
+				return;
+			}
+
+			this.dragEnterNodes = without( this.dragEnterNodes, Array.from( mutation.removedNodes ) );
+		} );
+	}
+
+	toggleDraggingOverDocument( event ) {
+		// Track nodes that have received a drag event. So long as nodes exist
+		// in the set, we can assume that an item is being dragged on the page.
+		if ( 'dragenter' === event.type && ! includes( this.dragEnterNodes, event.target ) ) {
+			this.dragEnterNodes.push( event.target );
+		} else if ( 'dragleave' === event.type ) {
+			this.dragEnterNodes = without( this.dragEnterNodes, event.target );
+		}
+
+		// In some contexts, it may be necessary to capture and redirect the
+		// drag event (e.g. atop an `iframe`). To accommodate this, you can
+		// create an instance of CustomEvent with the original event specified
+		// as the `detail` property.
+		//
+		// See: https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Creating_and_triggering_events
+		const detail = window.CustomEvent && event instanceof window.CustomEvent ? event.detail : event;
+
+		const { onVerifyValidTransfer } = this.props;
+		const isValidDrag = ! onVerifyValidTransfer || onVerifyValidTransfer( detail.dataTransfer );
+		const isDraggingOverDocument = isValidDrag && this.dragEnterNodes.length;
+
+		this.setState( {
+			isDraggingOverDocument: isDraggingOverDocument,
+			isDraggingOverElement: isDraggingOverDocument && this.isWithinZoneBounds( detail.clientX, detail.clientY ),
+		} );
+
+		if ( window.CustomEvent && event instanceof window.CustomEvent ) {
+			// For redirected CustomEvent instances, immediately remove window
+			// from tracked nodes since another "real" event will be triggered.
+			this.dragEnterNodes = without( this.dragEnterNodes, window );
+		}
+	}
+
+	preventDefault( event ) {
+		event.preventDefault();
+	}
+
+	isWithinZoneBounds( x, y ) {
```

Comment:
If a block is selected with a drop zone, and media is dragged over the editor, can we assume it's for this block? It may make the movement a bit easier.